zfsbootmenu/lib/zfsbootmenu-core.sh

```diff
@@ -2096,3 +2095,20 @@ mount_efivarfs() {
     mount -t efivarfs efivarfs "${efivar_location}" -o "${efivar_state}"
   fi
 }
+
+# arg1: zfs dataset name
+# returns: 0 if filesystem
+
+is_zfs_filesystem() {
+  local dataset
+
+  dataset="${1}"
+
+  if [ -z "${dataset}" ]; then
+    zerror "dataset undefined"
+    return 1
+  fi
+
+  [[ "$( zfs get -H -o value type "${dataset}" 2>&1 )" == "filesystem" ]] && return 0
```

Comment:
```suggestion
  zfs list -H -o name -t filesystem "${dataset}" 2>&1 >/dev/null && return 0
```
This seems to work as expected for filesystem, zvol, snapshot and nonexistent arguments, but should be tested a bit more thoroughly.

Reply:
I want it actually validating that it's a filesystem. zsnapshots can't operate on a zvol, snapshot, etc.

Reply:
What does this offer over a string comparison test - speed?

Reply:
It doesn't offer much, but doesn't rely on any program output except the error code when the type isn't as expected.